packages/client/src/pages/Game/engine/TDEngine.tsx

```diff
@@ -0,0 +1,770 @@
+    public waveCountdown: IWaveGenerator["waveCountdown"] = 0,
+    public isUICountdown = false,
+  ) {
+    this.waveCountdown = Math.floor(this.waveTimeoutBetweenWaves / 1000);
+  }
+
+  public repeatEnemy = (times: number) => {
+    let enemiesArray = [];
+    for (let iteration = 0; iteration < times; iteration++) {
+      // boss enemy
+      if (iteration === 10) {
+        enemiesArray.push(
+          new Enemy(this.engine, this.engine.enemySprites.boss!, {
+            width: 25,
+            height: 25,
+            spaceBetweenEnemies: 10,
+            speed:
+              0.3 +
+              this.waveParams.speedCoefficient * this.waveParams.currentWave,
+            bounty:
+              10 +
+              this.waveParams.enemyBountyCoefficient *
+                this.waveParams.currentWave,
+            hp:
+              1500 +
+              this.waveParams.hpCoefficient * this.waveParams.currentWave,
+          }),
+        );
+      } else if (iteration % 10 === 0) {
+        // slow enemy
+        enemiesArray.push(
+          new Enemy(this.engine, this.engine.enemySprites.slow!, {
+            width: 25,
+            height: 25,
+            spaceBetweenEnemies: 25,
+            speed:
+              0.4 +
+              this.waveParams.speedCoefficient * this.waveParams.currentWave,
+            bounty:
+              3 +
+              this.waveParams.enemyBountyCoefficient *
+                this.waveParams.currentWave,
+            hp:
+              250 + this.waveParams.hpCoefficient * this.waveParams.currentWave,
+          }),
+        );
+      } else if (iteration % 5 === 0) {
+        // fast enemy
+        enemiesArray.push(
+          new Enemy(this.engine, this.engine.enemySprites.fast!, {
+            width: 20,
+            height: 20,
+            spaceBetweenEnemies: 35,
+            speed:
+              1 +
+              this.waveParams.speedCoefficient * this.waveParams.currentWave,
+            bounty:
+              3 +
+              this.waveParams.enemyBountyCoefficient *
+                this.waveParams.currentWave,
+            hp:
+              50 + this.waveParams.hpCoefficient * this.waveParams.currentWave,
+          }),
+        );
+      } else {
+        // regular enemy
+        enemiesArray.push(
+          new Enemy(this.engine, this.engine.enemySprites.levelOne!, {
+            width: 20,
+            height: 20,
+            spaceBetweenEnemies: 35,
+            speed:
+              0.65 +
+              this.waveParams.speedCoefficient * this.waveParams.currentWave,
+            bounty:
+              1 +
+              this.waveParams.enemyBountyCoefficient *
+                this.waveParams.currentWave,
+            hp:
+              100 + this.waveParams.hpCoefficient * this.waveParams.currentWave,
+          }),
+        );
+      }
+    }
+    return enemiesArray;
+  };
+
+  public init() {
+    if (!this.isInitialized) {
```

Comment:
Sometimes it's better to negate the condition to make nested conditional statements easier to read:
```suggestion
    if (this.isInitialized) return;
```

Reply:
Skipped, isInitialized is only runonce flag, so WaveGenerator is not being executed once when isInitialized === false